calyx-opt/src/passes/compile_invoke.rs

```diff
@@ -147,39 +151,82 @@ impl CompileInvoke {
                 concrete_cell.borrow().ports.len()
             );
 
-            // Mapping from canonical names of the ports of the ref cell to the
-            // new port defined on the signature of the component. This has name of ref cell, not arg cell
-            let Some(comp_ports) = self.port_names.get(&inv_comp) else {
-                unreachable!("component `{}` invoked but not already visited by the pass", inv_comp)
+            // comp_ports is mapping from canonical names of the ports of the ref cell to the
+            // new port defined on the signature of the higher level component.
+            // i.e. ref_reg.in -> ref_reg_in
+            // These have name of ref cell, not the cell passed in as an arugment
+            let Some(comp_ports) = self.port_names.get(&inv_comp_id) else {
+                unreachable!("component `{}` invoked but not already visited by the pass", inv_comp_id)
             };
 
+            //contains the newly added ports that result from ref cells removal/dump_ports
+            let new_comp_ports = comp_ports
+                .values()
+                .map(|p| p.borrow().name)
+                .collect::<HashSet<_>>();
+
+            // tracks ports used in assigments of the invoked component
+            let mut used_ports: HashSet<ir::Id> = HashSet::new();
+            // If `is_none()` then the invoked comp should be a primitive , and
+            // we assume that all ports are used
+            if let Some(invoked_comp) = invoked_comp {
+                invoked_comp.iter_assignments(|a| {
+                    for port in a.iter_ports() {
+                        used_ports.insert(port.borrow().name);
+                    }
+                });
+                invoked_comp.iter_static_assignments(|a| {
+                    for port in a.iter_ports() {
+                        used_ports.insert(port.borrow().name);
+                    }
+                });
+            } else {
+                used_ports.clone_from(&new_comp_ports);
+            }
+
+            let to_assign: HashSet<&ir::Id> =
+                new_comp_ports.intersection(&used_ports).collect();
+
```

Comment:
This is the main new logic that tracks which ports are used in an invoked cell. Further down ports that aren't used are skipped over